RV32M1 SoC and VEGAboard support

mbolivar opened the pull request:
Edit: removed DNM tag, the series this depended on was merged

This adds support for the OpenISA VEGAboard's RI5CY processor, as well as board support for VEGAboard itself.

It does add a new vendor HAL; it's BSD-3-Clause licensed.

Commit:
cmake: gcc.cmake: allow users to influence CMAKE_REQUIRED_FLAGS

A GCC-based toolchain may require additional, toolchain-specific
values in CMAKE_REQUIRED_FLAGS to perform compiler checks properly,
but gcc.cmake clobbers any values the user provides.

Preserve them instead, allowing users to give their own compiler
checking flags at generation time.

The details for the particular issue that inspired this are described
in https://github.com/pulp-platform/pulpino/issues/240.

Signed-off-by: Marti Bolivar <marti@foundries.io>

mike-scott committed:
pinmux: RV32M1: introduce pinmux driver / DT bindings

Add a pinmux driver.

Signed-off-by: Michael Scott <mike@foundries.io>
Signed-off-by: Marti Bolivar <marti@foundries.io>

mike-scott committed:
serial: RV32M1: introduce lpuart driver / DT bindings

Add a UART driver.

Signed-off-by: Michael Scott <mike@foundries.io>
Signed-off-by: Marti Bolivar <marti@foundries.io>

mike-scott committed:
clock_control: RV32M1: introduce PCC driver / DT bindings

Add a Peripheral Clock Controller (PCC) driver. This gates and ungates
clocks to various peripherals on the SoC.

Signed-off-by: Michael Scott <mike@foundries.io>
Signed-off-by: Marti Bolivar <marti@foundries.io>

mike-scott committed:
gpio: RV32M1: introduce gpio driver / DT bindings

Add a GPIO driver.

Signed-off-by: Michael Scott <mike@foundries.io>

Commit:
arch: riscv32: support CONFIG_WDOG_INIT

Borrow from an Arm Cortex-M convention where each Kconfig.soc can
define a 'config WDOG_INIT' that does watchdog initialization early in
the boot process if that SoC needs it.

Some SoCs have watchdogs that are enabled by default and need to be
turned off during reset handling (to be re-enabled if necessary by a
Zephyr watchdog driver).

Signed-off-by: Marti Bolivar <marti@foundries.io>

Commit:
drivers: interrupt_controller: multi-level improvements

Some extensions to the multi-level interrupt controller are required
to support SoCs with more than four level 2 interrupt "aggregators".

Extend existing support to allow at most 8 level 2 or level 3
aggregators. Use Kconfig macro templates to cut down on boilerplate.
Try to clarify some aspects of the Kconfig help while we're at it, and
change the type of options which count things or are table offsets
from "hex" to "int", so that the generated .config is easier to read.

Finally, make some improvements to gen_isr_tables.py while we are
here. In particular, move some assignments around to cut down on
duplicated work, don't check for symbols we know must exist, and
improve the debug logging output's readability.

Signed-off-by: Marti Bolivar <marti@foundries.io>

Commit:
soc: riscv32: add RV32M1 SoC as openisa_rv32m1

The OpenISA RV32M1 SoC has four CPU cores. Two of these are RISC-V
32-bit cores, which are named "RI5CY" and "ZERO-RISCY". (The other two
cores are ARM Cortex-M0+ and -M4.) This patch adds basic SoC
enablement for the RISC-V cores:

- basic dtsi, to be extended as additional drivers are added
- SoC definition in soc/riscv32/openisa_rv32m1 for RI5CY / ZERO-RISCY
- system timer driver for RI5CY, based on LPTMR0 peripheral

The timer driver will be generalized a bit soon once proper
multi-level interrupt support is available.

Emphasis is on supporting the RI5CY core as the more capable of the
two; the ZERO-RISCY SoC definitions are a good starting point, but
additional work setting up a dtsi and initial drivers is needed to
support that core.

Signed-off-by: Marti Bolivar <marti@foundries.io>
Signed-off-by: Michael Scott <mike@foundries.io>

Commit:
interrupt_controller: RV32M1: add intmux driver / DT bindings

Add a level 2 interrupt controller for the RV32M1 SoC. This uses the
INTMUX peripheral.

As a first customer, convert the timer driver over to using this,
adding nodes for the LPTMR peripherals. This lets users select the
timer instance they want to use, and what intmux channel they want to
route its interrupt to, using DT overlays.

Signed-off-by: Marti Bolivar <marti@foundries.io>
Signed-off-by: Mike Scott <mike@foundries.io>

Commit:
boards: riscv32: add rv32m1_vega

This board directory supports the RV32M1 Vega board when targeting
a RISC-V CPU core on the main SoC.

Currently, only RI5CY support is provided via the rv32m1_vega_ri5cy
board name.

Signed-off-by: Marti Bolivar <marti@foundries.io>
Signed-off-by: Michael Scott <mike@foundries.io>

Commit:
ext: hal: add openisa/vega_sdk_riscv

This provides a HAL for the OpenISA RV32M1 SoC.

Origin: open-isa-rv32m1 GitHub organization
URL: https://github.com/open-isa-rv32m1/rv32m1_sdk_riscv
Revision: 365b1060f0947d5250c07b3eebdbc9e54cd0246e
Maintained-by: External
License: BSD-3-Clause

Signed-off-by: Marti Bolivar <marti@foundries.io>

mbolivar commented:
Fixed some of the style problems. Working on getting a license and copyright for openocd.cfg.

Many of the checkpatch suggestions would break if applied, so will have to be ignored.

SebastianBoe commented on arch/common/gen_isr_tables.py:
```diff
@@ -153,13 +155,31 @@ def write_source_file(fp, vt, swt, intlist):
+    if "CONFIG_2ND_LVL_ISR_TBL_OFFSET" in syms:
+        level2_offset = syms["CONFIG_2ND_LVL_ISR_TBL_OFFSET"]
+    else:
+        level2_offset = None
```
level2_offset.get("CONFIG_2ND_LVL_ISR_TBL_OFFSET", None)

SebastianBoe commented on arch/riscv32/core/fatal.c:
```diff
@@ -199,8 +199,9 @@ FUNC_NORETURN void _Fault(const NANO_ESF *esf)
-
+#ifdef CONFIG_PRINTK
 	printk("Exception cause %s (%d)\n", cause_str(mcause), (int)mcause);
+#endif
```
Users don't expect to see ifdef's around CONFIG_PRINTK.

We should instead drop the #ifdef CONFIG_PRINTK around mcause. This has no performance impact, and makes this function easier to read.

mbolivar replied:
This was merged as part of #11200. I will add a commit making this change to this series.

SebastianBoe commented on cmake/compiler/gcc.cmake:
```diff
@@ -138,5 +138,13 @@ endif()
+#
+# Appending onto any existing values lets users specify
+# toolchain-specific flags at generation time.
+set(CMAKE_REQUIRED_FLAGS ${CMAKE_REQUIRED_FLAGS} -nostartfiles -nostdlib ${isystem_include_flags} -Wl,--unresolved-symbols=ignore-in-object-files)
```
list(APPEND

A part from that, this patch looks sane.

mbolivar replied:
Much better.

SebastianBoe commented on ext/hal/openisa/vega_sdk_riscv/Kconfig:
```diff
@@ -0,0 +1,7 @@
+#
+# SPDX-License-Identifier: Apache-2.0
+
+menuconfig HAS_VEGA_SDK_HAL
```
"HAS_" options should not have any direct effect on the build when enabled.

It seems you have no need for a "HAS_" option, so you can simply rename this to "VEGA_SDK_HAL".

SebastianBoe commented on soc/riscv32/openisa_rv32m1/CMakeLists.txt:
```diff
@@ -0,0 +1,22 @@
+    # The RI5CY core has some ISA extensions that are best supported
+    # using the Pulpino toolchain. Use of a generic toolchain is
+    # untested and left to future work.
+    message(FATAL_ERROR "generic toolchain is not supported")
```
This assertion should be moved into Kconfig.

Also, it doesn't make sense to have code after a FATAL_ERROR.

SebastianBoe commented on soc/riscv32/openisa_rv32m1/linker.ld:
```diff
@@ -0,0 +1,219 @@
+
+#if defined(CONFIG_CUSTOM_RODATA_LD) || defined(CONFIG_CUSTOM_RWDATA_LD)
+#error "Custom linker data not supported"
+#endif
```
It is OK to have these assertions here. But they should redundantly exist in Kconfig as well.

SebastianBoe reviewed (changes requested):
Generally this looks very good.

But see misc. review comments.

dbkinder commented on arch/riscv32/Kconfig:
```diff
@@ -20,13 +20,49 @@ menu "RISCV32 Processor Options"
+	    See gen_offset.h for more details.
+
+	  The generic architecture IRQ wrapper will also call
+	  __soc_save_context and __soc_restore_context routines at
```
I think these underscores are causing the doc failure (WARNING: Inline emphasis start-string without end-string), so change to ``\_\_soc_save_context and \_\_soc_restore_context``

dbkinder commented on arch/riscv32/Kconfig:
```diff
@@ -20,13 +20,49 @@ menu "RISCV32 Processor Options"
+
+	  .. code-block:: c
+
+	     void __soc_save_context(soc_esf_t *state);
```
These underscores are OK (in a code block)

dbkinder commented on boards/riscv32/rv32m1_vega/doc/rv32m1_vega.rst:
```diff
@@ -0,0 +1,777 @@
+each for the RV32M1 SoC\ [#toolchain_openocd]_.
+
+Option 1 (Recommended): Prebuilt Toolchain and OpenOCD
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```
Our project conventions are to use dashes ``----`` for this level of heading

dbkinder commented on boards/riscv32/rv32m1_vega/doc/rv32m1_vega.rst:
```diff
@@ -0,0 +1,777 @@
+  this won't work with Zephyr's build system.
+
+Option 2: Building Toolchain and OpenOCD From Source
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```
ditto (dashes here too)

mbolivar replied:
Are you sure? I get:

```
/home/mbolivar/src/zephyr/doc/_build/rst/doc/boards/riscv32/rv32m1_vega/doc/rv32m1_vega.rst:310: WARNING: Title level inconsistent:

Option 1 (Recommended): Prebuilt Toolchain and OpenOCD
------------------------------------------------------
/home/mbolivar/src/zephyr/doc/_build/rst/doc/boards/riscv32/rv32m1_vega/doc/rv32m1_vega.rst:372: WARNING: Title level inconsistent:

Option 2: Building Toolchain and OpenOCD From Source
----------------------------------------------------
```

dbkinder commented on boards/riscv32/rv32m1_vega/doc/rv32m1_vega.rst:
```diff
@@ -0,0 +1,777 @@
+
+Ubuntu 18.04 users need to install these additional dependencies::
+
+  sudo apt-get install autoconf automake autotools-dev curl libmpc-dev libmpfr-dev libgmp-dev gawk build-essential bison flex texinfo gperf libtool patchutils bc zlib1g-dev libusb-1.0-0-dev libudev1 libudev-dev g++
```
Recommend breaking this long line into multiple lines (ending with a ``\``)

dbkinder reviewed (changes requested):
Nice documentation, thanks!

mbolivar commented on drivers/timer/rv32m1_lptmr_timer.c:
```diff
@@ -0,0 +1,168 @@
+ */
+
+#ifdef CONFIG_TICKLESS_IDLE
+#error "Tickless idle not yet implemented for RV32M1 LPTMR timer"
```
Ah, of course! Thanks.

Commit:
arch: riscv32: follow-on tweaks

A couple of follow-on patches suggested after previous RISCV32 arch
changes were merged.

Tweak some help in arch/riscv32/kconfig to better work with the RST
docs.

Take out all the CONFIG_PRINTK ifdeffery in fatal.c. The cause_str()
routine should get compiled out if PRINTK=n anyway.

Signed-off-by: Marti Bolivar <marti@foundries.io>

mbolivar commented:
@dbkinder @SebastianBoe comments addressed, thanks! Let's see what CI says.

alistair23 commented:
Can this be implemented following the RISC-V spec and not have a dependency on custom pulp toolchains?

With this diff I am able to build it using upstream GCC (not yet tested on hardware)

```
diff --git a/soc/riscv32/openisa_rv32m1/CMakeLists.txt b/soc/riscv32/openisa_rv32m1/CMakeLists.txt
index e4b5a8d535..d79089d3c4 100644
--- a/soc/riscv32/openisa_rv32m1/CMakeLists.txt
+++ b/soc/riscv32/openisa_rv32m1/CMakeLists.txt
@@ -3,7 +3,7 @@
 # SPDX-License-Identifier: Apache-2.0
 
 if(CONFIG_SOC_OPENISA_RV32M1_RI5CY)
-  zephyr_compile_options(-march=rv32imcxpulpv2)
+  zephyr_compile_options(-march=rv32imfdc)
 elseif(CONFIG_SOC_OPENISA_RV32M1_ZERO_RISCY)
   zephyr_compile_options(-march=rv32imc)
 endif()
```

alistair23 commented on boards/riscv32/rv32m1_vega/doc/rv32m1_vega.rst:
```diff
@@ -0,0 +1,782 @@
+   - the SEGGER `J-Link Software and Documentation Pack`_ software
+     installed
+
+   A JTAG dongle is not included with the board itself.
```
Is this really the only way? Anyone who wants to use this board needs to spend hundreds of dollars on a JTAG debugger?

mbolivar replied:
> The [open-isa website](http://open-isa.org/get-started/) implies that everything you need is included in the box. Is there no other way to use the board besides buying a specialised JTAG debugger?

There is no other way.

dbkinder commented on boards/riscv32/rv32m1_vega/doc/rv32m1_vega.rst:
```diff
@@ -0,0 +1,782 @@
+===   ========   ===========================================
+
+Additional Pins
+~~~~~~~~~~~~~~~
```
Please use our document conventions and use ``------`` here and below instead of ``~~~~~~`` for this heading level.

mbolivar commented:
@dbkinder sorry, neglected to mention, I've addressed your comments.

@alistair23: 

> With this diff I am able to build it using upstream GCC (not yet tested on hardware), if this could at least be described somewhere that would help others.

That's not really in my purview. I'm just a contractor and the recommendation I got from OpenISA is to use their toolchain. At any rate, I don't believe your patch is correct; specifically this part:

```
  zephyr_compile_options(-march=rv32imfdc)
```

If you check the reference manual, the core is an IMC core, but it's not documented to have any floating point support (the F and D extensions named in the above -march are single and double floating point support).

I've made this change:

```
diff --git a/soc/riscv32/openisa_rv32m1/CMakeLists.txt b/soc/riscv32/openisa_rv32m1/CMakeLists.txt
index e4b5a8d535..433ac77751 100644
--- a/soc/riscv32/openisa_rv32m1/CMakeLists.txt
+++ b/soc/riscv32/openisa_rv32m1/CMakeLists.txt
@@ -3,7 +3,11 @@
 # SPDX-License-Identifier: Apache-2.0
 
 if(CONFIG_SOC_OPENISA_RV32M1_RI5CY)
-  zephyr_compile_options(-march=rv32imcxpulpv2)
+  if (CONFIG_RISCV_GENERIC_TOOLCHAIN)
+    zephyr_compile_options(-march=rv32imc)
+  else()
+    zephyr_compile_options(-march=rv32imcxpulpv2)
+  endif()
 elseif(CONFIG_SOC_OPENISA_RV32M1_ZERO_RISCY)
   zephyr_compile_options(-march=rv32imc)
 endif()
```

That only uses the extensions specifically documented to be available. Since RISCV_GENERIC_TOOLCHAIN defaults to n for this soc (see soc/riscv32/openisa_rv32m1/Kconfig.defconfig), the pulp toolchain is chosen by default. You can set RISCV_GENERIC_TOOLCHAIN to y with a conf overlay to test out the standard toolchains if you want -- no vega-specific documentation needed here as this kconfig option is already part of zephyr.

alistair23 commented:
> @dbkinder sorry, neglected to mention, I've addressed your comments.
> 
> @alistair23:
> 
> > With this diff I am able to build it using upstream GCC (not yet tested on hardware), if this could at least be described somewhere that would help others.
> 
> That's not really in my purview. I'm just a contractor and the recommendation I got from OpenISA is to use their toolchain. At any rate, I don't believe your patch is correct; specifically this part:
> 
> ```
>   zephyr_compile_options(-march=rv32imfdc)
> ```
> 
> If you check the reference manual, the core is an IMC core, but it's not documented to have any floating point support (the F and D extensions named in the above -march are single and double floating point support).

You are right about the floating point. I just skimmed the documentation and saw that floating point support is optional and assumed that it was included. Thanks for catching that.

> 
> I've made this change:
> 
> ```
> diff --git a/soc/riscv32/openisa_rv32m1/CMakeLists.txt b/soc/riscv32/openisa_rv32m1/CMakeLists.txt
> index e4b5a8d535..433ac77751 100644
> --- a/soc/riscv32/openisa_rv32m1/CMakeLists.txt
> +++ b/soc/riscv32/openisa_rv32m1/CMakeLists.txt
> @@ -3,7 +3,11 @@
>  # SPDX-License-Identifier: Apache-2.0
>  
>  if(CONFIG_SOC_OPENISA_RV32M1_RI5CY)
> -  zephyr_compile_options(-march=rv32imcxpulpv2)
> +  if (CONFIG_RISCV_GENERIC_TOOLCHAIN)
> +    zephyr_compile_options(-march=rv32imc)
> +  else()
> +    zephyr_compile_options(-march=rv32imcxpulpv2)
> +  endif()
>  elseif(CONFIG_SOC_OPENISA_RV32M1_ZERO_RISCY)
>    zephyr_compile_options(-march=rv32imc)
>  endif()
> ```

This looks much better :+1: 

> 
> That only uses the extensions specifically documented to be available. Since RISCV_GENERIC_TOOLCHAIN defaults to n for this soc (see soc/riscv32/openisa_rv32m1/Kconfig.defconfig), the pulp toolchain is chosen by default. You can set RISCV_GENERIC_TOOLCHAIN to y with a conf overlay to test out the standard toolchains if you want -- no vega-specific documentation needed here as this kconfig option is already part of zephyr.

Ok, sounds good to me. I'm unable to test it on hardware but at least being able to build Zephyr with standard GCC is a good start.

dbkinder reviewed (approved):
doc changes LGTM, thanks.

mbolivar commented:
recheck

mbolivar commented:
retest

mbolivar commented:
Thanks for all the reviews, everyone!

The checkpatch issues should be ignored (its recommendations would break correctness), so it's just a license review now.

I've added the TSC label and asked for their review on the vendor HAL.

rgundi reviewed (commented):
The changes made to the multiple interrupt levels code look good. Neatly done. Thanks for doing this. This can be considered as my approval for this section of the PR.

galak commented:
Would be nice if one of the RISCV guys reviewed & approved before we merge.

galak commented:
I can't remember but is this one of those ports that we want the new RISC-V toolchain for?

mbolivar commented:
> Would be nice if one of the RISCV guys reviewed & approved before we merge.

@kgugala @pgielda @nategraff-sifive mind giving this a review then? You're listed in CODEOWNERS for {arch,soc,boards}/riscv32

nashif commented:
@mbolivar mind rebasing?

mbolivar commented:
> @kgugala @pgielda @nategraff-sifive mind giving this a review then? You're listed in CODEOWNERS for {arch,soc,boards}/riscv32

Ping? Sorry to bug you, but I'm under pressure to get this merged...

mbolivar commented:
Hah, never mind :)
